graveler/diff.go

```diff
@@ -0,0 +1,104 @@
+package graveler
+
+import (
+	"bytes"
+	"fmt"
+)
+
+type diffIterator struct {
+	left       ValueIterator
+	right      ValueIterator
+	leftNext   bool
+	rightNext  bool
+	currentVal *Diff
+	err        error
+}
+
+func NewDiffIterator(left ValueIterator, right ValueIterator) DiffIterator {
+	it := &diffIterator{left: left, right: right}
+	it.leftNext = it.left.Next()
+	it.rightNext = it.right.Next()
+	return it
+}
+
+func (d *diffIterator) compareKeys() int {
+	if d.leftNext && !d.rightNext {
+		return -1
+	}
+	if !d.leftNext && d.rightNext {
+		return 1
+	}
+	return bytes.Compare(d.left.Value().Key, d.right.Value().Key)
+}
+
+func (d *diffIterator) Next() bool {
+	if d.err != nil {
+		return false
+	}
+	for d.leftNext || d.rightNext {
+		if d.left.Err() != nil {
+			d.err = fmt.Errorf("failed in left tree: %w", d.left.Err())
+			return false
+		}
+		if d.right.Err() != nil {
+			d.err = fmt.Errorf("failed in right tree: %w", d.right.Err())
+			return false
+		}
+		comp := d.compareKeys()
+		switch comp {
+		case 0:
+			leftVal := d.left.Value()
+			rightVal := d.right.Value()
+			d.rightNext = d.right.Next()
+			d.leftNext = d.left.Next()
+			if !bytes.Equal(leftVal.Value.Identity, rightVal.Value.Identity) {
+				d.currentVal = &Diff{
+					Type:  DiffTypeChanged,
+					Key:   rightVal.Key,
+					Value: rightVal.Value,
+				}
+				return true
+			}
+		case -1:
+			d.currentVal = &Diff{
+				Type:  DiffTypeRemoved,
+				Key:   d.left.Value().Key,
+				Value: d.left.Value().Value,
+			}
+			d.leftNext = d.left.Next()
+			return true
+		default: // leftKey > rightKey
+			d.currentVal = &Diff{
+				Type:  DiffTypeAdded,
+				Key:   d.right.Value().Key,
+				Value: d.right.Value().Value,
+			}
+			d.rightNext = d.right.Next()
+			return true
+		}
+	}
+	d.currentVal = nil
```

Comment:
I think you do not take care of right reaching end while left still has entries. in that cas, all those entries are additions. IIUC you will not handle this case.